libsql-wal/src/segment/list.rs

```diff
@@ -0,0 +1,173 @@
+use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+use std::sync::Arc;
+
+use arc_swap::ArcSwapOption;
+use fst::{map::OpBuilder, Streamer};
+use libsql_sys::ffi::Sqlite3DbHeader;
+use memoffset::offset_of;
+use zerocopy::{AsBytes, FromZeroes};
+
+use crate::error::Result;
+use crate::fs::file::FileExt;
+use crate::segment::{sealed::SealedSegment, Frame};
+
+struct SegmentLink<F> {
+    segment: SealedSegment<F>,
+    next: ArcSwapOption<SegmentLink<F>>,
+}
+
+pub struct SegmentList<F> {
+    head: ArcSwapOption<SegmentLink<F>>,
+    len: AtomicUsize,
+    /// Whether the segment list is already being checkpointed
+    checkpointing: AtomicBool,
+}
+
+impl<F> Default for SegmentList<F> {
+    fn default() -> Self {
+        Self {
+            head: Default::default(),
+            len: Default::default(),
+            checkpointing: Default::default(),
+        }
+    }
+}
+
+impl<F> SegmentList<F> {
+    /// Prepend the list with the passed sealed segment
+    pub fn push_log(&self, segment: SealedSegment<F>) {
+        let segment = Arc::new(SegmentLink {
+            segment,
+            next: self.head.load().clone().into(),
+        });
+
+        self.head.swap(Some(segment));
```

Comment:
these two lines are not atomic, that is okay? i was expecting to see operation similar to CAS

Reply:
I don't understand your concern. What do you expect and why?

Reply:
Is following possible:

1. t1: Call to `push_log`, the first line gets executed `let segment = Arc::new...`. This `segment` points to the old head `h1`
2. t2: (Another thread), same call to `push_log`, this thread runs both the instructions, now the head points to the new head
3. t3: the first thread gets resumed, inserts the old `segment` object

(my rust is weak, so I was wondering if this can happen)